### PR TITLE
[SPARK-11119][SQL] cleanup for unsafe array and map

### PR DESCRIPTION
**Pull request by cloud-fan**
The purpose of this PR is to keep the unsafe format detail only inside the unsafe class itself, so when we use them(like use unsafe array in unsafe map, use unsafe array and map in columnar cache), we don't need to understand the format before use them.

change list:
- unsafe array's 4-bytes numElements header is now required(was optional), and become a part of unsafe array format.
- w.r.t the previous changing, the `sizeInBytes` of unsafe array now counts the 4-bytes header.
- unsafe map's format was `[numElements] [key array numBytes] [key array content(without numElements header)] [value array content(without numElements header)]` before, which is a little hacky as it makes unsafe array's header optional. I think saving 4 bytes is not a big deal, so the format is now: `[key array numBytes] [unsafe key array] [unsafe value array]`.
- w.r.t the previous changing, the `sizeInBytes` of unsafe map now counts both map's header and array's header.

### CODE REVIEW

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
cc @davies

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeMapData.java [hunk 1]
```diff
@@ -17,41 +17,75 @@
...
+   * @param sizeInBytes the size of this map's backing data, in bytes
+   */
+  public void pointTo(Object baseObject, long baseOffset, int sizeInBytes) {
+    // Read the numBytes of key array from the first 4 bytes.
+    final int keyArraySize = Platform.getInt(baseObject, baseOffset);
+    final int valueArraySize = sizeInBytes - keyArraySize - 4;
+    assert keyArraySize >= 0 : "keyArraySize (" + keyArraySize + ") should >= 0";
+    assert valueArraySize >= 0 : "valueArraySize (" + valueArraySize + ") should >= 0";
+
+    final UnsafeArrayData keyArray = new UnsafeArrayData();
```
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeMapData.java]: We do not need to create these two objections for each `pointTo`, they should be created in constructor.

**Comment by davies** [comment]
LGTM, left one minor comment.

**Comment by davies** [comment]
Merged into master, thanks!

**Pull request closed without merging**